src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs

```diff
@@ -52,50 +55,64 @@ private static void ClearCachedData()
+            var client = await RemoteHostClient.TryGetClientAsync(project, cancellationToken).ConfigureAwait(false);
 
-            var documentKeys = project.Documents.SelectAsArray(DocumentKey.ToDocumentKey);
-            var priorityDocumentKeys = priorityDocuments.SelectAsArray(DocumentKey.ToDocumentKey);
+            await foreach (var item in ConvertItemsAsync(
+                project.Solution, activeDocument, SearchCachedDocumentsWorkerAsync(), cancellationToken).ConfigureAwait(false))
+            {
+                yield return item;
+            }
 
-            var client = await RemoteHostClient.TryGetClientAsync(project, cancellationToken).ConfigureAwait(false);
-            if (client != null)
+            IAsyncEnumerable<RoslynNavigateToItem> SearchCachedDocumentsWorkerAsync()
             {
-                var callback = new NavigateToSearchServiceCallback(onItemFound);
-                await client.TryInvokeAsync<IRemoteNavigateToSearchService>(
-                    (service, callbackId, cancellationToken) =>
-                        service.SearchCachedDocumentsAsync(documentKeys, priorityDocumentKeys, searchPattern, kinds.ToImmutableArray(), callbackId, cancellationToken),
-                    callback, cancellationToken).ConfigureAwait(false);
+                var solution = project.Solution;
 
-                return;
-            }
+                var documentKeys = project.Documents.SelectAsArray(DocumentKey.ToDocumentKey);
+                var priorityDocumentKeys = priorityDocuments.SelectAsArray(DocumentKey.ToDocumentKey);
 
-            var storageService = solution.Services.GetPersistentStorageService();
-            await SearchCachedDocumentsInCurrentProcessAsync(
-                storageService, documentKeys, priorityDocumentKeys, searchPattern, kinds, onItemFound, cancellationToken).ConfigureAwait(false);
+                if (client != null)
+                {
+                    var channel = Channel.CreateUnbounded<RoslynNavigateToItem>();
+
+                    // Kick off the work to do the search in another thread.  That work will push the results into the
+                    // channel.  When the work finishes (for any reason, including cancellation), the channel will be 
+                    // completed.
+                    Task.Run(async () => await client.TryInvokeAsync<IRemoteNavigateToSearchService>(
+                            (service, callbackId, cancellationToken) =>
+                                service.SearchCachedDocumentsAsync(documentKeys, priorityDocumentKeys, searchPattern, kinds.ToImmutableArray(), callbackId, cancellationToken),
+                            new NavigateToSearchServiceCallback(channel), cancellationToken).ConfigureAwait(false), cancellationToken)
+                        .CompletesChannel(channel);
+
+                    return channel.Reader.ReadAllAsync(cancellationToken);
```

Comment:
Here's a pattern you'll see in the 4 places in AbstractNavigateToSearchService where we call into OOP.

1. We make a channel (for the results to be pushed into), we kick off a task to call into OOP to do the searching.
2. The searching task is given a callback which pushes into the channel.
3. The task 'Completes' the channel when it is down (that's how the reader knows there's no more)
4. The channel is read-from as an IAsyncEnumerable to return those values to the caller.

Reply:
I presume we want to remove this in favor of full async enumerable once the streamjsonrpc issue is resolved.  Should we have a tracking item on that?